File: src/prog/event/EventPayload.java
package prog.event;

/**
 * Type-safe payload for flight data events.
 * Replaces the untyped Map&lt;String, String&gt; for compile-time safety
 * and zero unnecessary String boxing of boolean/numeric values.
 *
 * Immutable — safe for cross-thread passing between Service and EDT.
 */
public final class EventPayload {
    public final String mapGrid;
    public final boolean fatalWarn;
    public final boolean radioAltValid;
    public final boolean isDowningFlap;
    public final String timeStr;
    public final boolean isJet;
    public final boolean engineCheckDone;
    /** Optimal compressor stage index (0-based). -1 indicates invalid/jet/single-stage. */
    public final int optimalCompressorStage;
    /** True when actual compressor stage doesn't match optimal (at full throttle). */
    public final boolean compressorStageMismatch;

    public EventPayload(String mapGrid, boolean fatalWarn, boolean radioAltValid,
                        boolean isDowningFlap, String timeStr, boolean isJet,
                        boolean engineCheckDone, int optimalCompressorStage,
                        boolean compressorStageMismatch) {
        this.mapGrid = mapGrid;
        this.fatalWarn = fatalWarn;
        this.radioAltValid = radioAltValid;
        this.isDowningFlap = isDowningFlap;
        this.timeStr = timeStr;
        this.isJet = isJet;
        this.engineCheckDone = engineCheckDone;
        this.optimalCompressorStage = optimalCompressorStage;
        this.compressorStageMismatch = compressorStageMismatch;
    }

    public static class Builder {
        private String mapGrid = "--";
        private boolean fatalWarn = false;
        private boolean radioAltValid = false;
        private boolean isDowningFlap = false;
        private String timeStr = "--:--";
        private boolean isJet = false;
        private boolean engineCheckDone = false;
        private int optimalCompressorStage = -1;
        private boolean compressorStageMismatch = false;

        public Builder mapGrid(String v) { this.mapGrid = v; return this; }
        public Builder fatalWarn(boolean v) { this.fatalWarn = v; return this; }
        public Builder radioAltValid(boolean v) { this.radioAltValid = v; return this; }
        public Builder isDowningFlap(boolean v) { this.isDowningFlap = v; return this; }
        public Builder timeStr(String v) { this.timeStr = v; return this; }
        public Builder isJet(boolean v) { this.isJet = v; return this; }
        public Builder engineCheckDone(boolean v) { this.engineCheckDone = v; return this; }
        public Builder optimalCompressorStage(int v) { this.optimalCompressorStage = v; return this; }
        public Builder compressorStageMismatch(boolean v) { this.compressorStageMismatch = v; return this; }

        public EventPayload build() {
            return new EventPayload(mapGrid, fatalWarn, radioAltValid,
                                    isDowningFlap, timeStr, isJet, engineCheckDone,
                                    optimalCompressorStage, compressorStageMismatch);
        }
    }

    public static Builder builder() { return new Builder(); }
}

File: rust/crates/vm-core/src/event/event_payload.rs
//! 对应 Java: `src/prog/event/EventPayload.java`
//! Type-safe payload for flight data events.
//! Replaces the untyped Map<String, String> for compile-time safety
//! and zero unnecessary String boxing of boolean/numeric values.
//!
//! Immutable — safe for cross-thread passing between Service and EDT.
//! PORT: Java final class + public final 字段 = Rust pub struct + pub 字段 (§0.7),
//! 不可变性由"构造后只经 &self 访问"保证 (Java 靠 final 引用)。
//! PORT: Java 无 equals 覆写 (引用等值); 此处 derive PartialEq 仅为测试基建
//! (fields.rs 先例), 不改变翻译逻辑。

/// Type-safe payload for flight data events. (radioAltValid 已删: 零消费方)
#[derive(Debug, Clone, PartialEq)]
pub struct EventPayload {
    pub map_grid: String,
    pub fatal_warn: bool,
    pub is_downing_flap: bool,
    pub time_str: String,
    pub is_jet: bool,
    pub engine_check_done: bool,
    /// Optimal compressor stage index (0-based). -1 indicates invalid/jet/single-stage.
    pub optimal_compressor_stage: i32,
    /// True when actual compressor stage doesn't match optimal (at full throttle).
    pub compressor_stage_mismatch: bool,
}

impl EventPayload {
    /// 对应 Java 公有构造器 `EventPayload(String, boolean, ..., boolean)`。
    // PORT: Java 保真 — 参数表逐个对应 Java 构造器形参, 不打包成结构体
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        map_grid: String,
        fatal_warn: bool,
        is_downing_flap: bool,
        time_str: String,
        is_jet: bool,
        engine_check_done: bool,
        optimal_compressor_stage: i32,
        compressor_stage_mismatch: bool,
    ) -> Self {
        EventPayload {
            map_grid,
            fatal_warn,
            is_downing_flap,
            time_str,
            is_jet,
            engine_check_done,
            optimal_compressor_stage,
            compressor_stage_mismatch,
        }
    }

    /// 对应 Java `public static Builder builder()`。
    pub fn builder() -> EventPayloadBuilder {
        // Java: return new Builder(); — 每次调用返回全新 Builder (字段取初始缺省值)
        EventPayloadBuilder::new()
    }
}

/// 对应 Java `EventPayload.Builder` (public static class)。
/// 缺省值逐字段对齐 Java 字段初始化器: mapGrid="--", timeStr="--:--",
/// optimalCompressorStage=-1, 其余 false。
#[derive(Debug, Clone)]
pub struct EventPayloadBuilder {
    map_grid: String,
    fatal_warn: bool,
    is_downing_flap: bool,
    time_str: String,
    is_jet: bool,
    engine_check_done: bool,
    optimal_compressor_stage: i32,
    compressor_stage_mismatch: bool,
}

impl Default for EventPayloadBuilder {
    /// 补齐 Rust 惯用 Default (语义同 `Builder::new()`, Java 侧无此概念)
    fn default() -> Self {
        Self::new()
    }
}

impl EventPayloadBuilder {
    /// 对应 Java Builder 默认构造器 `new Builder()` (public)。
    pub fn new() -> Self {
        EventPayloadBuilder {
            map_grid: "--".to_string(),
            fatal_warn: false,
            is_downing_flap: false,
            time_str: "--:--".to_string(),
            is_jet: false,
            engine_check_done: false,
            optimal_compressor_stage: -1,
            compressor_stage_mismatch: false,
        }
    }

    /// Java: `Builder mapGrid(String v) { this.mapGrid = v; return this; }`
    pub fn map_grid(mut self, v: String) -> Self {
        self.map_grid = v;
        self
    }

    pub fn fatal_warn(mut self, v: bool) -> Self {
        self.fatal_warn = v;
        self
    }

    pub fn is_downing_flap(mut self, v: bool) -> Self {
        self.is_downing_flap = v;
        self
    }

    pub fn time_str(mut self, v: String) -> Self {
        self.time_str = v;
        self
    }

    pub fn is_jet(mut self, v: bool) -> Self {
        self.is_jet = v;
        self
    }

    pub fn engine_check_done(mut self, v: bool) -> Self {
        self.engine_check_done = v;
        self
    }

    pub fn optimal_compressor_stage(mut self, v: i32) -> Self {
        self.optimal_compressor_stage = v;
        self
    }

    pub fn compressor_stage_mismatch(mut self, v: bool) -> Self {
        self.compressor_stage_mismatch = v;
        self
    }

    /// 对应 Java `EventPayload build()` — 只读字段组包, 不修改 Builder。
    pub fn build(&self) -> EventPayload {
        EventPayload::new(
            self.map_grid.clone(),
            self.fatal_warn,
            self.is_downing_flap,
            self.time_str.clone(),
            self.is_jet,
            self.engine_check_done,
            self.optimal_compressor_stage,
            self.compressor_stage_mismatch,
        )
    }
}

#[cfg(test)]
mod tests;
